File: PythonCodes/Tab06_01.py
#%%
"""
Normal density recovery using the COS method- Convergence exercise
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

def COSDensity(cf,x,N,a,b):
    i = np.complex(0.0,1.0) #assigning i=sqrt(-1)
    k = np.linspace(0,N-1,N)
    u = np.zeros([1,N])
    u = k * np.pi / (b-a)
        
    # F_k coefficients

    F_k    = 2.0 / (b - a) * np.real(cf(u) * np.exp(-i * u * a));
    F_k[0] = F_k[0] * 0.5; # adjustment for the first term
    
    # Final calculation

    f_X = np.matmul(F_k , np.cos(np.outer(u, x - a )))
        
    # We output only the first row

    return f_X
    
def mainCalculation():
    i = np.complex(0.0, 1.0) #assigning i=sqrt(-1)
    
    # Setting for the COS method 

    a = -10
    b = 10
    
    # Define the range for the expansion terms

    N = [2**x for x in range(2,7,1)]
    
    # Parameter setting for normal distribution

    mu = 0
    sigma = 1 
        
    # Define characteristic function for the normal distribution

    cF = lambda u : np.exp(i * mu * u - 0.5 * np.power(sigma,2.0) * np.power(u,2.0));
    
    # Define the domain for density

    x = np.linspace(-5.0,5,1000)
    f_XExact = st.norm.pdf(x,mu,sigma)
    
    plt.figure(1)
    plt.grid()
    plt.xlabel("x")
    plt.ylabel("$f_X(x)$")
    for n in N:
        f_X = COSDensity(cF,x,n,a,b)
        error = np.max(np.abs(f_X-f_XExact))
        print("For {0} expanansion terms the error is {1}".format(n,error))
        
        plt.plot(x,f_X)
        plt.legend(["N=%.0f"%N[0],"N=%.0f"%N[1],"N=%.0f"%N[2],"N=%.0f"%N[3],"N=%.0f"%N[4]])
    
    
mainCalculation()

File: MatlabCodes/Tab06_01.m
function COS_Normal_Density_Convergence
close all;clc;
format long
i = complex(0,1); %assigning i=sqrt(-1)

% The COS method settings

a = -10; % truncation domain
b =  10;

% Characteristic function parameters

mu     = 0;    sigma  = 1;

% Domain for the density f(x)

x = linspace(-5,5,1000);

% Characteristic function  

cf  =  @(u) exp(i * mu * u - 0.5 * sigma^2 * u.^2);

% Iterate over different numbers of expansion terms

Ngrid = 2.^(2:1:5);
idx = 1;
error = zeros(length(Ngrid),1);
figure(1); hold on; grid on;xlabel('x');ylabel('f_X(x)')
legend_txt= cell(length(Ngrid),1);
for N=Ngrid

    % Density from the COS method

    f_X = COSDensity(cf,x,N,a,b);

    % Error

    error(idx) = max(normpdf(x)-f_X);
    plot(x,f_X)
    legend_txt{idx} = sprintf('N=%i',N);
    idx = idx + 1;
end
legend(legend_txt)

function f_X = COSDensity(cf,x,N,a,b)
i = complex(0,1); % Assigning i=sqrt(-1)
k = 0:N-1; 
u = k * pi / (b - a);

% F_k coefficients

F_k    = 2 / (b - a) * real(cf(u) .* exp(-i * u * a));
F_k(1) = F_k(1) * 0.5; % adjustment for the first term

% Final calculation

f_X = F_k * cos(u' * (x - a));
